py/torch_tensorrt/dynamo/_engine_caching.py

```diff
@@ -0,0 +1,268 @@
+import copy
+import logging
+import os
+import pickle
+import shutil
+import sys
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Optional, Tuple, cast
+
+import torch
+from torch._inductor.codecache import FxGraphCachePickler
+from torch.fx.experimental.proxy_tensor import maybe_disable_fake_tensor_mode
+
+_LOGGER: logging.Logger = logging.getLogger(__name__)
+
+
+class BaseEngineCache(ABC):
+
+    @abstractmethod
+    def __init__(
+        self,
+        *args: Any,
+        **kwargs: Any,
+    ) -> None:
+        pass
+
+    @staticmethod
+    def get_hash(gm: torch.fx.GraphModule) -> str:
+        """Get the hash value of the GraphModule
+
+        Args:
+            gm (torch.fx.GraphModule): GraphModule to hash
+
+        Returns:
+            str: hash value of the GraphModule
+        """
+        # parameters are set to 0
+        with maybe_disable_fake_tensor_mode():
+            new_gm = copy.deepcopy(gm)
+            for name, param in new_gm.named_parameters():
+                param.data.zero_()
+
+            hash_val = cast(str, FxGraphCachePickler.get_hash(new_gm))
+
+        return hash_val
+
+    @abstractmethod
+    def save(
+        self,
+        hash: str,
+        serialized_engine: bytes,
+        input_names: List[str],
+        output_names: List[str],
+        weight_name_map: Optional[Dict[str, Any]] = None,
+    ) -> bool:
+        """Save the serialized engine to hard disk
+
+        Args:
+            hash (str): hash value of the GraphModule
+            serialized_engine (bytes): serialized TRT engine
+            input_names (List[str]): input names of TRT engine
+            output_names (List[str]): output names of TRT engine
+            weight_name_map (Optional[Dict[str, Any]]): weight name map for refitting
+
+        Returns:
+            bool: whether the serialized engine is saved successfully
+        """
+        pass
+
+    @abstractmethod
+    def load(
+        self, hash: str
+    ) -> Tuple[Optional[bytes], List[str], List[str], Optional[Dict[str, Any]]]:
+        """Load the serialized engine from hard disk
+
+        Args:
+            hash (str): hash value of the GraphModule
+
+        Returns:
+            Sequence[Optional[bytes], List[str], List[str], Optional[Dict[str, Any]]]: serialized engine, input names, output names, weight name map
```

Comment:
You need to define exactly what a cache hit looks like and how a cache would signal that there isn't a hit.

Reply:
If we just give the cache a blob then its easy its just this function returns `bytes | None`